admin/services/pin-generation.service.js

```diff
@@ -88,19 +92,37 @@ pinGenerationService.isValid = async (p) => {
 /**
  * Generate pupils pins
  * @param pupilsList
+ * @param dfeNumber
  * @returns {Array}
  */
-pinGenerationService.updatePupilPins = async (pupilsList) => {
-  const ids = Object.values(pupilsList || null)
+pinGenerationService.updatePupilPins = async (pupilsList, dfeNumber) => {
+  let ids = Object.values(pupilsList || null)
+  ids = ids.map(i => parseInt(i))
   const pupils = await pupilDataService.sqlFindByIds(ids)
-  pupils.forEach(async pupil => {
+  pupils.forEach(pupil => {
     if (!pinValidator.isActivePin(pupil.pin, pupil.pinExpiresAt)) {
       pupil.pin = pinGenerationService.generatePupilPin()
       pupil.pinExpiresAt = pinExpiryTime()
     }
   })
   const data = pupils.map(p => ({ id: p.id, pin: p.pin, pinExpiresAt: p.pinExpiresAt }))
-  return pupilDataService.sqlUpdatePinsBatch(data)
+  try {
+    await pupilDataService.sqlUpdatePinsBatch(data)
+  } catch (error) {
+    // Handle duplicate pins
+    if (error.number === 2601 &&
+      pinGenerationService.pinSubmissionAttempts < pinGenerationService.pinSubmissionMaxAttempts) {
```

Comment:
let's say the first attempt fails.  How many more attempts will this logic make?